File: zigzag-conversion/zigzag-conversion.py
#leetcode - zigzag conversion - medium

class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        result = [[] for i in range(numRows)]
        i = 0
        write = 0
        while i < len(s):
            while write < numRows and i < len(s):
                print('down', write, s[i])
                result[write].append(s[i])
                write += 1
                i += 1  
            write = numRows - 2
            while write >= 0 and i < len(s):
                print('up', write, s[i])
                result[write].append(s[i])
                write -= 1
                i += 1
            write = 1
                
        collect = []
        for word in result:
            collect.extend(word)
        return ''.join(collect)

File: zigzag-conversion/zigzag-conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
    if (numRows == 1) {
        return s
    };
    const collect = new Array(numRows).fill(null).map(() => []);
    let i = 0;
    let row = 0;
    while (i < s.length) {
        while (i < s.length && row < numRows) {
            collect[row].push(s[i]);
            i += 1;
            row += 1
        };
        row = numRows - 2;
        while (i < s.length && row >= 0) {
            collect[row].push(s[i]);
            i += 1;
            row -= 1
        };
        row = 1
    };
    let result = collect.flat().join('');
    return result
};